source/common/router/scoped_rds.cc

```diff
@@ -0,0 +1,272 @@
+#include "common/router/scoped_rds.h"
+
+#include <memory>
+
+#include "envoy/admin/v2alpha/config_dump.pb.h"
+#include "envoy/api/v2/srds.pb.validate.h"
+
+#include "common/common/assert.h"
+#include "common/config/subscription_factory.h"
+
+// Types are deeply nested under Envoy::Config::ConfigProvider; use 'using-directives' across all
+// ConfigProvider related types for consistency.
+using Envoy::Config::ConfigProvider;
+using Envoy::Config::ConfigProviderInstanceType;
+using Envoy::Config::ConfigProviderManager;
+using Envoy::Config::ConfigProviderPtr;
+
+namespace Envoy {
+namespace Router {
+
+ConfigProviderPtr ScopedRoutesConfigProviderUtil::maybeCreate(
+    const envoy::config::filter::network::http_connection_manager::v2::HttpConnectionManager&
+        config,
+    Server::Configuration::FactoryContext& factory_context, const std::string& stat_prefix,
+    ConfigProviderManager& scoped_routes_config_provider_manager) {
+  if (config.route_specifier_case() != envoy::config::filter::network::http_connection_manager::v2::
+                                           HttpConnectionManager::kScopedRoutes) {
+    return nullptr;
+  }
+
+  switch (config.scoped_routes().config_specifier_case()) {
+  case envoy::config::filter::network::http_connection_manager::v2::ScopedRoutes::
+      kScopedRouteConfigurationsList: {
+    const envoy::config::filter::network::http_connection_manager::v2::
+        ScopedRouteConfigurationsList& scoped_route_list =
+            config.scoped_routes().scoped_route_configurations_list();
+    std::vector<std::unique_ptr<const Protobuf::Message>> config_protos(
+        scoped_route_list.scoped_route_configurations().size());
+    for (const auto& it : scoped_route_list.scoped_route_configurations()) {
+      Protobuf::Message* clone = it.New();
+      clone->CopyFrom(it);
+      config_protos.push_back(std::unique_ptr<const Protobuf::Message>(clone));
+    }
```

Comment:
Do we need to copy, or could we pass the `RepeatedFieldPtr` intro `createStaticConfigProvider` somehow?

Reply:
To keep the `Envoy::Config::ConfigProviderManager::createStaticConfigProvider` interface generic, the container passed into the function must contain `Protobuf::Message` pointers. Switching to `RepeatedFieldPtr<Protobuf::Message>` would still trigger a copy of each item in the container.

Reply:
OK, maybe use `std::copy` or `std::transform` then.

Reply:
I removed this code from the PR since it is not yet used. I will reintroduce it in the next PR with modifications based on the feedback.